### ISSUE TITLE
win32eventreactor doesn't notice TCP connection is lost in certain (hopefully rare) cases

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5233]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5233|
|Type|defect|
|Created|2011-08-14 14:04:56Z|
|Branch|https://github.com/twisted/twisted/tree/win32er-close-5233-2|

http://twistedmatrix.com/trac/browser/branches/abortConnection-78-7/twisted/internet/test/test_tcp.py?rev=32436#L1413 -- `test_resumeProducingThrows` -- demonstrates the win32eventreactor problem: it doesn't notice a lost TCP connection. The problem doesn't appear to have anything to do with the `abortConnection` call; the test still fails if that line is commented out.

Protocol A connects over loopback to Protocol B. Protocol A's `resumeProducing` gets called from its transport's `doWrite`; `resumeProducing` throws an exception. At this point we expect the following to happen:

   1. Exception thrown from resumeProducing is logged, and Protocol A has its `connectionLost` called.
   2. Protocol B, which is connected to A over TCP, should also get its `connectionLost` called.

While item 1 happens, item 2 does not. As far as I can tell, `MsgWaitForMultipleObjects` never notices that B has lost its connection.

Only workaround I found: making the socket close with SO_LINGER set to 1,0, i.e. force RST instead of FIN. In this case MWFMO does notice the connection was lost.

Attachments:

* [WSAEnumNetworkEvents.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d96/d965a463ac183af4828302b14522bb3371cec49f/a3c8950d1371e563f114e3e7511e060212538adb.patch) (22610 bytes) - added by zseil on 2011-10-23 19:22:03Z - standalone WSAEnumNetworkEvents

<details><summary>Searchable metadata</summary>

```
trac-id__5233 5233
type__defect defect
reporter__itamar itamar
priority__low low
milestone__ 
branch__branches_win32er_close_5233_2 branches/win32er-close-5233-2
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__windows windows
time__1313330696000000 1313330696000000
changetime__1330307539000000 1330307539000000
version__None None
owner__exarkun exarkun
cc__zooko@...
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to @exarkun|
|-|-|

[#5285](https://github.com/twisted/twisted/issues/5285) was a duplicate of this.  There's a self-contained example attached there.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [32695]) Branching to 'win32er-close-5233'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

This is probably fixed in the branch.  However, it's not tested.  The test will be annoying to write.  Maybe I'll give it a try soon.  I also observed that this issue is probably also the cause of glib2/gtk2reactor flakiness on Windows.  glib2 I/O notification propagates the one-shot nature of FD_CLOSE notification through to us, and we don't do anything special to handle it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Oops, we don't use glib2/gtk2 on Windows, we use `PortableGtkReactor`.  So nevermind about those bits.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [32909]) Branching to 'win32er-close-5233-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun removed owner|
|-|-|

Okay.  I think I added a decent test, and made the fix really work in all cases.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/win32er-close-5233-2)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst set owner to @itamarst|
|-|-|

I'll try to look at this tomorrow. For now, two notes:

1. The test that was disabled in [#7153](https://github.com/twisted/twisted/issues/7153) should be re-enabled (if it's not already, haven't looked in detail at branch yet).

2. We should open new ticket, and add the potential explanation, for pygtk issue noted in branch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

One issue that the build results pointed out is that WSAEnumNetworkEvents is somewhat new - it was added in late 2010.  And several of our build slaves have a version of pywin32 which is old enough to not have it (disabling win32eventreactor entirely).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst set owner to @exarkun|
|-|-|

As far as availability of WSAEnumNetworkEvents, we can require people to upgrade pywin32, try to recreate the functionality with ctypes, or just fall back to the buggy behavior if old version of pywin32 is installed (in which case we should emit warning when the module is imported).

Given that going the ctypes route requires a bunch of extra work and tests, I think my preferred option would be requiring an upgrade. Mysterious non-closing sockets are not cool, and it's not like Windows users want to stick to particular versions of libraries distributed with the OS as you would on Linux.

I see item 1 above has been addressed - excellent.

2. Item 2 above - add the guess about reason for pygtk issues to appropriate ticket.

3. Any reason you're not using sets?

4. The way in which an event leaves the closingReading pair of data structures is very... implicit. It seems like it work though, so I guess it's OK, but it seems tricksy enough that I'd want a test that it's not actually leaking objects.

Otherwise, looks good -- great job debugging a mysterious issue.

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)<a name="note_10"></a>|@zseil commented|
|-|-|

I've attached standalone WSAEnumNetworkEvents module if you'd prefer that over requiring users to upgrade the pywin32 package. The code is basically the same as in pywin32, there are also some tests.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun removed owner|
|-|-|

  1. r32969 makes pywin32 215 the explicitly required version.
  1. Filed [#5333](https://github.com/twisted/twisted/issues/5333) for pygtk/windows/close
  1. Not really.  I'm weakly attached to the symmetry allowed by using a dict for `_closedAndReading` since there is no `WeakSet` to use for `_closedAndNotReading`, only a `WeakKeyDictionary`.
  1. I added a test in r32972.

As far as having our own `WSAEnumNetworkEvents` extension module... I'm not extremely opposed, but it means figuring out a behavior on PyPy (maybe not immediately, I don't think pywin32 works there yet), maintaining some more C, etc.  I think in general we shouldn't worry about bumping our Windows dependencies, though, particularly if it lets us skip maintaining more C code.  In any case, though, thanks very much for making that option easier for us, zseil. :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

And [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/win32er-close-5233-2).

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)<a name="note_13"></a>|@zseil set owner to @exarkun|
|-|-|

Nice! Only one small issue, at http://twistedmatrix.com/trac/browser/branches/win32er-close-5233-2/twisted/internet/win32eventreactor.py?rev=32969#L269, you should also catch `ValueError` which is raised when `fd.fileno()` returns an invalid handle. This will happen e.g. for `SerialPort`s.

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)<a name="note_14"></a>|@zseil commented|
|-|-|

Ups, my last comment was wrong, `pywintypes.error` is raised if `fd` is not a socket but looks like a valid handle.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

Does the branch in its current form break serial port support on Windows?  I'm guessing so, but I have no way to test the serial port code.

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)<a name="note_16"></a>|@zseil commented|
|-|-|

Replying to [exarkun](#note_15):
> Does the branch in its current form break serial port support on Windows?  I'm guessing so, but I have no way to test the serial port code.

Yes, the reactor explodes if a `SerialPort` is added to it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [32981]) Handle pywintypes.error from WSAEnumNetworkEvents, to keep serial ports working

refs [#5233](https://github.com/twisted/twisted/issues/5233)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun removed owner|
|-|-|

Okay, handling pywintypes.error now.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

[final build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/win32er-close-5233-2). reviewing.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph set owner to @exarkun<br>@glyph set status to `new`|
|-|-|

I think I've reviewed this as well as I am able (since I don't have a physical serial port on Windows: is it possible to create one with VMWare somehow?) 

This looks pretty good.  Excellent diagnostic work, I must say, given how obscure this condition is.

The one problem is that I really don't like the idea of the surprise explosion when upgrading Twisted on an existing installation without upgrading pywin32 first.  Let's have a warning before we have IOCP and win32er just stop working completely.  It's certainly better to very, very occasionally lose a `connectionLost` than to be completely unable to use the reactor.  Just because we don't have to be _as_ concerned about win32 users upgrading dependencies doesn't mean we don't have to be concerned at _all_; there is still the issue of upgrading development environments, synchronizing build systems, etc.

That said, I mostly agree with exarkun's observations about not wanting to maintain that much more C.  Thanks, zseil, for implementing that mammoth function wrapper extension module, but I suspect that it will make more work for us than it saves, in the long run :).

I think we should get handle an `ImportError` on attempting to import `WSAEnumNetworkEvents`, and not depend on the results from `GetFileVersionInfo`.  It's always possible that they'll change their versioning scheme in the future and this number will be wrong.  Although it may be useful to call `GetFileVersionInfo` in the `except` block to provide the build number of the currently installed version so the user has some idea of how far behind they are.  We should change the error to a warning, and the warning should say _where you download pywin32 from_, so that this doesn't become a FAQ for frustrated Windows users :).

Then, since we already have to handle the case where `WSAEnumNetworkEvents` doesn't actually _work_ (the serial port stuff) we can just have a stub `WSAEnumNetworkEvents` that raises `TypeError` (and warns you again, maybe).  This will still cause the test suite to fail, and that's fine; the failures can totally tell you to upgrade :).

Despite the volumes of prose here, I suspect the actual diff for making this change will be tiny, so if you agree with this plan then please just go ahead and land it after upgrading the buildbots.

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)<a name="note_22"></a>|@zseil commented|
|-|-|

I just noticed another thing, if you'd call `WSAEnumNetworkEvents` only for `fd`'s that are in `self._reads`, you could disconnect the `fd` when `WSAEnumNetworkEvents` raises an error and unskip the `twisted.internet.test.test_fdset.ReactorFDSetTestsBuilder.test_negativeOneFileDescriptor` test.

This would also be more backwards compatible, `fd.fileno()` was previously never called for `fd`s added directly with `reactor.addEvent()`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun removed owner|
|-|-|

Thanks for the review, and special thanks for the additional comments, zseil.  It's actually quite important to not call `fileno` for non-socket fds, since that method is not part of the interface for those objects.

I added an explicit fileno() == -1 check and got rid of the explicit pywin32 version check.

[Latest build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/win32er-close-5233-2)

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)|@zseil set owner to @zseil|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)<a name="note_25"></a>|@zseil set owner to @exarkun|
|-|-|

I think this is ready now. Both iocp and win32er reactors can now be imported with older pywin32 versions, `test_resumeProducingThrows` expectedly fails in that configuration. While there is no new test checking that fileno() doesn't get called for descriptors added with addEvent, existing tests would fail if that would happen now that there is an explicit fileno() call (test_win32events.Listener doesn't have a fileno method). Buildbot failures either look unrelated (hardy32-py2.5-select) or should get fixed once the branch is merged to trunk.

Thanks for working on this, even when Windows is not your favorite platform. Please merge!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun set status to `closed`|
|-|-|

(In [33049]) Merge win32er-close-5233-2

Author: exarkun
Reviewer: itamar, zseil, glyph
Fixes: [#5233](https://github.com/twisted/twisted/issues/5233)

win32eventreactor is now more careful in its handling of socket close
notifications.  Where previously it could miss a close notification
entirely if it arrived at the same time as the last chunk of bytes on
a connection, it will now handle that case and properly disconnect
the transport and protocol.  However, this functionality requires

pywin32 215 or later; it will be disabled with a warning if an older
version of pywin32 is installed.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_27"></a>|@zooko commented|
|-|-|

You may enjoy (?) the consternation that this bugfix causes me over on the Tahoe-LAFS trac: https://tahoe-lafs.org/trac/tahoe-lafs/ticket/1681
